<a href="https://colab.research.google.com/github/EVA5-Stars/EVA5/blob/master/S6/EVA5_Submission_Session_6_in_progress.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Objective take best 5th code, and run below versions for 25 epochs and report findings:

>> 1. with L1 + BN
2. with L2 + BN
3. with L1 and L2 with BN
4. with GBN
5. with L1 and L2 with GBN


* *You need to write a single loop or iterator to iterate through these conditions.*

* Draw ONE graph to show the validation accuracy curves for all 5 jobs above. This graph must have proper legends and it should be clear what we are looking at.

* Draw ONE graph to show the loss change curves for all 5 jobs above. This graph must have proper legends and it should be clear what we are looking at.

* Find any 25 misclassified images (combined into single image) for "with GBN" model. You should be using the saved model from the above jobs.  You MUST show the actual and predicted class names.

Submitted By: 
* Gaurav Makkar
* Mohit Bhandari
* Suman Debnath
* Sriram Iyengar




## Import `pytorch` libraries and modules:

In [1]:
from __future__ import print_function
import torch
#import nueral net model in the torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
#torchvision is the libraray in torch that has many features
#and we are using datasets and transforms from this library 
from torchvision import datasets, transforms

In [2]:
!rm -rf EVA5
!git clone https://github.com/EVA5-Stars/EVA5.git

Cloning into 'EVA5'...
remote: Enumerating objects: 157, done.
remote: Counting objects: 100% (157/157), done.
remote: Compressing objects: 100% (110/110), done.
remote: Total 157 (delta 89), reused 91 (delta 43), pack-reused 0
Receiving objects: 100% (157/157), 620.15 KiB | 888.00 KiB/s, done.
Resolving deltas: 100% (89/89), done.


In [3]:
%matplotlib inline
import matplotlib.pyplot as plt

In [4]:
from google.colab import drive
drive.mount('/content/gdrive')

!ls /content/gdrive/My\ Drive/Colab\ Notebooks/EVA5/S6

Mounted at /content/gdrive
EVA5_Submission_Session_6_inprogress.ipynb
EVA5:Submission-Session_6_Model-BatchNormalization.pt
EVA5:Submission-Session_6_Model-GhostBatchNormalization.pt
L1EVA5:Submission-Session_6_Model-BatchNormalization.pt
L1EVA5:Submission-Session_6_Model-GhostBatchNormalization.pt


In [5]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")

## Project Library:

*   Ghost Batch Normalization
*   Misclassified image collection
*   Load the testing and training data with configurabe *batch_size*
*   Plot the Loss and Accuracy Graph
*   Define the structure of the network
*   Setup to create model and summary
*   Load the testing and training data with configurabe *batch_size*
*   Define test and train functions
*   Create a function to *initiate_training* for all conditions





**Ghost Batch Normalization**: 

A technique originally developed for training with very large batch sizes across many accelerators. 

In [6]:
#from EVA5.S6.BatchNorm_GhostBatchNorm import GhostBatchNorm
#from EVA5.S6.BatchNorm_GhostBatchNorm import BatchNorm

**Collecting Misclassified Images**

In [7]:
#from S6.S6_plot_misclassfied_image_finder import misclassified_image_finder              
def misclassified_image_finder(model, model_path, device, train_loader, image_num, msg):
    
    data_iter = iter(test_loader)
    figure = plt.figure()

    plt.title('Misclassified Images: With {}'.format(msg))
   
    for _i in range(image_num):
          data, target = data_iter.next()

          model.load_state_dict(torch.load(model_path)) 
          model.eval()

          data, target = data.to(device), target.to(device)

          output = model(data)
          pred = output.argmax(dim=1, keepdim=True) 
          leng=len(pred)
          for a in range(leng):
              if(pred[a]!=target[a]):                  
                  plt.subplot(5,5,_i+1)
                  plt.axis('off')
                  plt.imshow(data[a].cpu().numpy().squeeze(),cmap='gray_r')
                  

Plot the Loss and Accuracy Graph

In [8]:
from EVA5.S6.S6_plot_stat_loss_acc import plot_stat

Define the structure of the network

In [9]:
from EVA5.S6.S6_Model_structure_create import Net

Setup to create model and summary

In [10]:
from EVA5.S6.S6_Model_structure_create import create_model

Load the testing and training data with configurabe *batch_size*

In [11]:
from EVA5.S6.S6_data_loader import init_train_test_loader

Define test and train functions

In [12]:
from EVA5.S6.S6_train_test_function import train
from EVA5.S6.S6_train_test_function import test   


Create a function to *initiate_training* for all conditions

In [13]:
from torch.optim.lr_scheduler import StepLR
from EVA5.S6.S6_init_training import init_training   


 Initialize train and test loader with Batch Size:128


Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:469: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


## Run: Initialization and trainings

Create the model with normal BatchNormalization

In [14]:
model = create_model(device,gbatnor=False)
# Set the Epoch
EPOCHS = 25
model_name_bn = "EVA5:Submission-Session_6_Model-BatchNormalization.pt"
save_path_bn = F"/content/gdrive/My Drive/Colab Notebooks/EVA5/S6/{model_name_bn}"
torch.save(model.state_dict(), save_path_bn)
print("Model Saved: {}".format(model_name_bn))

Create the instance of the Net class with GBN = False

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
            Conv2d-4           [-1, 16, 24, 24]           1,152
              ReLU-5           [-1, 16, 24, 24]               0
       BatchNorm2d-6           [-1, 16, 24, 24]              32
         MaxPool2d-7           [-1, 16, 12, 12]               0
            Conv2d-8            [-1, 8, 12, 12]             128
              ReLU-9            [-1, 8, 12, 12]               0
      BatchNorm2d-10            [-1, 8, 12, 12]              16
           Conv2d-11           [-1, 16, 10, 10]           1,152
             ReLU-12           [-1, 16, 10, 10]               0
      BatchNorm2d-13           [-1, 16, 10, 10] 

Create the model with Ghost BatchNormalization

In [15]:
model1 = create_model(device, gbatnor=True)
model_name_gbn = "EVA5:Submission-Session_6_Model-GhostBatchNormalization.pt"
save_path_gbn = F"/content/gdrive/My Drive/Colab Notebooks/EVA5/S6/{model_name_gbn}"
torch.save(model1.state_dict(), save_path_gbn)
print("Model Saved: {}".format(model_name_gbn))

Create the instance of the Net class with GBN = True
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
    GhostBatchNorm-3            [-1, 8, 26, 26]              16
            Conv2d-4           [-1, 16, 24, 24]           1,152
              ReLU-5           [-1, 16, 24, 24]               0
    GhostBatchNorm-6           [-1, 16, 24, 24]              32
         MaxPool2d-7           [-1, 16, 12, 12]               0
            Conv2d-8            [-1, 8, 12, 12]             128
              ReLU-9            [-1, 8, 12, 12]               0
   GhostBatchNorm-10            [-1, 8, 12, 12]              16
           Conv2d-11           [-1, 16, 10, 10]           1,152
             ReLU-12           [-1, 16, 10, 10]               0
   GhostBatchNorm-13           [-1, 16, 10, 10]   

1. Cases


In [ ]:
# Total Loss & Accuracy
total_train_losses = []
total_train_accuracies = []

total_test_losses = []
total_test_accuracies = []
#####
train_loader, test_loader = init_train_test_loader()
# with L1 and BN
Case1=["L1 and BN",model, device, train_loader, EPOCHS, False, 1e-5, False]
# with L2 and BN
Case2=["L2 and BN",model, device, train_loader, EPOCHS,False, None, True]
# with L1, L2 and BN
Case3=["L1, L2 and BN",model, device, train_loader, EPOCHS,  True, 1e-5,True]
# with GBN
Case4=["GBN",model1, device, train_loader, EPOCHS, False, 1e-5,False]
# with L1, L2 and GBN
Case5=["L1, L2 and GBN",model1, device, train_loader, EPOCHS, True, 1e-5, True]

Cases=[Case1,Case2,Case3,Case4,Case5]
for case in Cases:
  # Loss train & test
  train_losses = []
  test_losses = []  
  # Accuracy train & test
  train_acc = []
  test_acc = []
  case_name,model, device, train_loader, epochs, step_lr, l1_lambda, l2_en=case
  print(" Below output is for Case",case_name)
  #print(*case)
  init_training(model, device, train_loader, epochs,train_losses,train_acc,test_losses,test_acc, step_lr=step_lr, l1_lambda=l1_lambda, l2_en=l2_en)
  total_train_losses.extend([train_losses])
  total_train_accuracies.extend([train_acc])
  total_test_losses.extend([test_losses])
  total_test_accuracies.extend([test_acc])
  print(total_test_losses)
  print(total_test_accuracies)

Loss=2.351052761077881 Batch_id=0 Accuracy=8.59:   0%|          | 1/469 [00:00<01:16,  6.13it/s]


 Initialize train and test loader with Batch Size:128
 Below output is for Case L1 and BN


Loss=0.23027123510837555 Batch_id=468 Accuracy=86.80: 100%|██████████| 469/469 [00:55<00:00,  8.48it/s]



 Epoch 0, lr 0.01


Loss=0.27114158868789673 Batch_id=0 Accuracy=96.09:   0%|          | 1/469 [00:00<00:56,  8.28it/s]


Test set: Average loss: 0.2329, Accuracy: 9655/10000 (96.55%)



Loss=0.24084778130054474 Batch_id=468 Accuracy=96.63: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 1, lr 0.01


Loss=0.13123252987861633 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<00:53,  8.68it/s]


Test set: Average loss: 0.1260, Accuracy: 9743/10000 (97.43%)



Loss=0.14170759916305542 Batch_id=468 Accuracy=97.46: 100%|██████████| 469/469 [00:55<00:00,  8.51it/s]



 Epoch 2, lr 0.01


Loss=0.1604025959968567 Batch_id=0 Accuracy=96.09:   0%|          | 1/469 [00:00<00:55,  8.42it/s]


Test set: Average loss: 0.0943, Accuracy: 9806/10000 (98.06%)



Loss=0.09604119509458542 Batch_id=468 Accuracy=97.72: 100%|██████████| 469/469 [00:54<00:00,  8.55it/s]



 Epoch 3, lr 0.01


Loss=0.06593607366085052 Batch_id=0 Accuracy=100.00:   0%|          | 1/469 [00:00<01:00,  7.68it/s]


Test set: Average loss: 0.0680, Accuracy: 9869/10000 (98.69%)



Loss=0.060051701962947845 Batch_id=468 Accuracy=97.99: 100%|██████████| 469/469 [00:54<00:00,  8.53it/s]



 Epoch 4, lr 0.01


Loss=0.09984823316335678 Batch_id=0 Accuracy=97.66:   0%|          | 1/469 [00:00<00:53,  8.79it/s]


Test set: Average loss: 0.0631, Accuracy: 9869/10000 (98.69%)



Loss=0.14010289311408997 Batch_id=468 Accuracy=98.12: 100%|██████████| 469/469 [00:54<00:00,  8.55it/s]



 Epoch 5, lr 0.01


Loss=0.08990460634231567 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<00:53,  8.70it/s]


Test set: Average loss: 0.0516, Accuracy: 9881/10000 (98.81%)



Loss=0.07136587798595428 Batch_id=468 Accuracy=98.29: 100%|██████████| 469/469 [00:54<00:00,  8.55it/s]



 Epoch 6, lr 0.01


Loss=0.08571170270442963 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<00:55,  8.48it/s]


Test set: Average loss: 0.0485, Accuracy: 9884/10000 (98.84%)



Loss=0.060759492218494415 Batch_id=468 Accuracy=98.38: 100%|██████████| 469/469 [00:54<00:00,  8.55it/s]



 Epoch 7, lr 0.01


Loss=0.07581890374422073 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<01:04,  7.28it/s]


Test set: Average loss: 0.0477, Accuracy: 9879/10000 (98.79%)



Loss=0.05851052328944206 Batch_id=468 Accuracy=98.41: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 8, lr 0.01


Loss=0.041724156588315964 Batch_id=0 Accuracy=100.00:   0%|          | 1/469 [00:00<00:54,  8.63it/s]


Test set: Average loss: 0.0457, Accuracy: 9889/10000 (98.89%)



Loss=0.06356308609247208 Batch_id=468 Accuracy=98.50: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 9, lr 0.01


Loss=0.09158387035131454 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<00:54,  8.60it/s]


Test set: Average loss: 0.0497, Accuracy: 9873/10000 (98.73%)



Loss=0.06593872606754303 Batch_id=468 Accuracy=98.59: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 10, lr 0.01


Loss=0.0888138934969902 Batch_id=0 Accuracy=96.09:   0%|          | 1/469 [00:00<00:54,  8.63it/s]


Test set: Average loss: 0.0427, Accuracy: 9880/10000 (98.80%)



Loss=0.11823265254497528 Batch_id=468 Accuracy=98.61: 100%|██████████| 469/469 [00:54<00:00,  8.56it/s]



 Epoch 11, lr 0.01


Loss=0.0730455294251442 Batch_id=0 Accuracy=97.66:   0%|          | 1/469 [00:00<00:58,  7.96it/s]


Test set: Average loss: 0.0373, Accuracy: 9904/10000 (99.04%)



Loss=0.03918289393186569 Batch_id=468 Accuracy=98.69: 100%|██████████| 469/469 [00:55<00:00,  8.53it/s]



 Epoch 12, lr 0.01


Loss=0.0464884489774704 Batch_id=0 Accuracy=99.22:   0%|          | 1/469 [00:00<00:54,  8.52it/s]


Test set: Average loss: 0.0395, Accuracy: 9898/10000 (98.98%)



Loss=0.04914826899766922 Batch_id=468 Accuracy=98.77: 100%|██████████| 469/469 [00:55<00:00,  8.50it/s]



 Epoch 13, lr 0.01


Loss=0.04965530335903168 Batch_id=0 Accuracy=99.22:   0%|          | 1/469 [00:00<00:53,  8.77it/s]


Test set: Average loss: 0.0376, Accuracy: 9903/10000 (99.03%)



Loss=0.12339839339256287 Batch_id=468 Accuracy=98.72: 100%|██████████| 469/469 [00:55<00:00,  8.52it/s]



 Epoch 14, lr 0.01


Loss=0.04056987538933754 Batch_id=0 Accuracy=99.22:   0%|          | 1/469 [00:00<00:55,  8.45it/s]


Test set: Average loss: 0.0316, Accuracy: 9921/10000 (99.21%)



Loss=0.09335332363843918 Batch_id=468 Accuracy=98.83: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 15, lr 0.01


Loss=0.038052696734666824 Batch_id=0 Accuracy=100.00:   0%|          | 1/469 [00:00<00:53,  8.82it/s]


Test set: Average loss: 0.0338, Accuracy: 9898/10000 (98.98%)



Loss=0.035032618790864944 Batch_id=468 Accuracy=98.76: 100%|██████████| 469/469 [00:55<00:00,  8.52it/s]



 Epoch 16, lr 0.01


Loss=0.09986825287342072 Batch_id=0 Accuracy=96.88:   0%|          | 1/469 [00:00<00:58,  7.99it/s]


Test set: Average loss: 0.0335, Accuracy: 9911/10000 (99.11%)



Loss=0.09555476903915405 Batch_id=468 Accuracy=98.87: 100%|██████████| 469/469 [00:55<00:00,  8.51it/s]



 Epoch 17, lr 0.01


Loss=0.06500353664159775 Batch_id=0 Accuracy=97.66:   0%|          | 1/469 [00:00<00:53,  8.68it/s]


Test set: Average loss: 0.0311, Accuracy: 9917/10000 (99.17%)



Loss=0.07444456219673157 Batch_id=468 Accuracy=98.84: 100%|██████████| 469/469 [00:55<00:00,  8.50it/s]



 Epoch 18, lr 0.01


Loss=0.03540825843811035 Batch_id=0 Accuracy=99.22:   0%|          | 1/469 [00:00<00:58,  8.02it/s]


Test set: Average loss: 0.0304, Accuracy: 9919/10000 (99.19%)



Loss=0.025846993550658226 Batch_id=468 Accuracy=98.98: 100%|██████████| 469/469 [00:54<00:00,  8.54it/s]



 Epoch 19, lr 0.01


Loss=0.05375789478421211 Batch_id=0 Accuracy=97.66:   0%|          | 1/469 [00:00<00:54,  8.66it/s]


Test set: Average loss: 0.0287, Accuracy: 9924/10000 (99.24%)



Loss=0.042164940387010574 Batch_id=468 Accuracy=98.88: 100%|██████████| 469/469 [00:54<00:00,  8.55it/s]



 Epoch 20, lr 0.01


Loss=0.04369354620575905 Batch_id=0 Accuracy=98.44:   0%|          | 1/469 [00:00<00:57,  8.12it/s]


Test set: Average loss: 0.0283, Accuracy: 9922/10000 (99.22%)



Loss=0.03698575496673584 Batch_id=189 Accuracy=99.05:  41%|████      | 190/469 [00:22<00:32,  8.63it/s]

In [ ]:
#Plot for Test Losses for Different Cases
plot_stat(total_test_losses, 'Loss')

In [ ]:
# Plot for Test accuracy for Different Cases
print(total_test_accuracies)
plot_stat(total_test_accuracies, 'Acc')

In [ ]:
misclassified_image_finder(model, save_path_gbn, device, train_loader, EPOCHS, "Ghost Batch Normalization")

## Summary:

A simple DNN with Total params=7,612 was created and trained for 25 epochs with L1, L2 Regularization, Batch Normalization and Ghost Batch Normalization. 

Following are the accuracy observed:

> 1. with L1 + BN => **99.29%**
2. with L2 + BN => **99.31%**
3. with L1 and L2 with BN => **99.47%**
4. with GBN => **98.9%**
5. with L1 and L2 with GBN => **99.33%**

Reference:
* GBN:  https://github.com/apple/ml-cifar-10-faster/blob/master/utils.py

